contrib/babelfishpg_tsql/src/collation.c

```diff
@@ -505,40 +506,63 @@ Datum remove_accents_internal(PG_FUNCTION_ARGS)
 	}
 
 	len_uinput = icu_to_uchar(&utf16_input, input_str, strlen(input_str));
-	pfree(input_str);
 
 	limit = len_uinput;
 	/* 
-	 * set the capacity to limit * MAX_BYTES_PER_CHAR if it is less than INT32_MAX
-	 * else set it to INT32_MAX as capacity is of int32_t datatype so it can 
-	 * have maximum INT32_MAX value
+	 * set the capacity (In UChar terms) to limit * MAX_BYTES_PER_CHAR if it is less than INT32_MAX
+	 * else set it to INT32_MAX as capacity is of int32_t datatype so it can have maximum INT32_MAX
+	 * value which would be equivalent to 2GB UChar points and 2GB * sizeof(UChar) in byte terms.
+	 * XXX: It is assumed that this capacity should handle almost all the general input strings.
 	 */
 	capacity = (limit < (PG_INT32_MAX / MAX_BYTES_PER_CHAR)) ? (limit * MAX_BYTES_PER_CHAR) : PG_INT32_MAX;
 
+	/*
+	 * utrans_transUChars will modify input string in place so ensure that it has enough capacity to store
+	 * transformed string.
+	 */
+	utf16_res = (UChar *) palloc0(capacity * sizeof(UChar));
```

Comment:
palloc limit seems to be
```
#define MaxAllocSize	((Size) 0x3fffffff) /* 1 gigabyte - 1 */

#define AllocSizeIsValid(size)	((Size) (size) <= MaxAllocSize)
```
capacity * sizeof(UChar) can go upto 4GB.

Reply:
We can also probably change this ```Min((len_uinput + 1) * sizeof(UChar), capacity * sizeof(UChar))``` to ```(len_uinput + 1) * sizeof(UChar)```
Because ```len_uinput + 1``` will never exceed the limit of palloc, since ```icu_to_uchar``` tries to palloc the same length.

Reply:
This is good point. With the latest revision, we are limiting the size of input that this function can process. For longer term, we will try to chunk the input string into smaller parts, process it and concat back the final string.